README.md
# activity-tools

This project contains building blocks that can be used to create ActivityPub applications.

**Project status**: Experimental, work in progress

## Documentation

Feel free to browse the code, or read the generated [doumentation](https://nsg.github.io/activity-tools/activity_tools.html). If you prefer an example instead I have included a [run.py](https://github.com/nsg/activity-tools/blob/master/run.py) who runs a small sample instance.

## Install

Install it from [PyPI](https://pypi.org/project/activity-tools/). Versions after 1.0.0 will follow semantic versioning with a `major.minor.patch` where breaking changes should only occur with a major version bump.

```
pip install activity-tools
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.0.x

The `0.0.x` releases are experimental releases with incomplete code, functionality and have an unstable interface. Do not expect any working and reliable code.


run.py
"""

This is an example application that uses activity-tools with FastAPI to create a
simple demo API that you can browse and follow from an Mastodon server.

If you have installed activity-tools via pip, change the imports from
"src.activity_tools" to activity_tools".

Start the application with:
$ uvicorn run:app --host 0.0.0.0 --workers 2

"""

import os
import json
import requests

from fastapi import FastAPI, Request
from fastapi.responses import RedirectResponse, JSONResponse, FileResponse

from src.activity_tools.objects import Actor, WrapActivityStreamsObject, Follow, Undo, Accept
from src.activity_tools.misc import PublicKey, WebFinger
from src.activity_tools.headers import ContentTypes, verify_signature, make_signature
from src.activity_tools.inbox import Inbox
from src.activity_tools.crypto import RSAKey

DOMAIN = os.getenv("DOMAIN", "example.com")
KEY = RSAKey("/tmp/key.pem")

app = FastAPI(
    title="ActivityPub Example Application",
    description="Let's see what I can do with a few lines of Python!",
    version="0.1.0"
)

@app.get("/", include_in_schema=False)
def redirect_to_docs():
    response = RedirectResponse(url='/docs')
    return response

@app.get("/.well-known/webfinger")
def webfinger(resource: str):
    content = WebFinger(DOMAIN, resource).run()
    return JSONResponse(content=content, headers=ContentTypes.jrd)

@app.get("/users/{username}")
def users(username: str):

    actor = Actor()
    actor.create(DOMAIN, username, KEY.public_key)
    actor.followers = None
    actor.following = None
    actor.outbox = None

    return JSONResponse(content=actor.run(), headers=ContentTypes.activity)

@app.get("/users/{username}/key")
def users(username: str):
    public_key = KEY.public_key.decode()
    object = WrapActivityStreamsObject(PublicKey(DOMAIN, username, public_key))
    print(object.run())
    return JSONResponse(content=object.run(), headers=ContentTypes.activity)

@app.post('/users/{username}/inbox')
async def inbox(username: str, request: Request):

    body = json.loads(await request.body())

    if not verify_signature(body, request.headers.items(), request.url.path):
        content = { "message": "Invalid signature" }
        return JSONResponse(
            content=content,
            headers=ContentTypes.activity,
            status_code=401
        )

    inbox = Inbox(body)
    object = inbox.parse()

    if type(object) == Follow:
        follow = object
        accept = Accept(follow)

        respond_to_url = follow.actor.inbox
        public_key_url = follow.object.public_key["id"]
        signature = make_signature(respond_to_url, accept.run(), public_key_url)

        r = requests.post(
            object.actor.inbox,
            data=json.dumps(accept.run()),
            headers={ **ContentTypes.activity, **signature }
        )

        print(r.status_code, r.content)

    elif type(object) == Undo:
        print("We got a undo request!")


src/actor.py
from .misc import ImageAsset, PublicKey, Tags, Attachment

class Actor:

    def __init__(self, domain, username, public_key) -> None:
        self.domain = domain
        self.username = username
        self.public_key = public_key

        self.id = f"https://{self.domain}/users/{self.username}"
        self.type = "Person"

        self.inbox = f"https://{self.domain}/users/{self.username}/inbox"
        self.outbox = f"https://{self.domain}/users/{self.username}/outbox"

        self.following = f"https://{domain}/users/{username}/following"
        self.followers = f"https://{domain}/users/{username}/followers"

        self.discoverable = False
        self.summary = ""
        self.published = "1523-06-06T10:00:00Z"

        self.name = f"{self.username.capitalize()}"
        self.preferred_username = f"{self.username}"

        self.icon_url = None
        self.image_url = None
        self.manually_approves_followers = None
        self.attachment = Attachment()
        self.tag = Tags(self.domain)

    def add_property_value(self, name, value):
        self.attachment.add_property_value(name, value)

    def add_emoji(self, name, url):
        self.tag.add_emoji(name, url)

    def run(self) -> dict:
        required_document = {
            "@context": [
                "https://www.w3.org/ns/activitystreams",
                "https://w3id.org/security/v1",
            ],
            "id": self.id,
            "type": self.type,
            "inbox": self.inbox,
            "outbox": self.outbox,
            "following": self.following,
            "followers": self.followers,
            "discoverable": self.discoverable,
            "summary": self.summary,
            "published": self.published,
            "name": self.name,
            "preferredUsername": self.preferred_username,
            "attachment": self.attachment.run(),
            "tag": self.tag.run(),
            "publicKey": PublicKey(self.domain, self.username, "public key").run(),
        }

        extra_values = {}

        if self.icon_url:
            extra_values["icon"] = ImageAsset(self.icon_url)

        if self.icon_url:
            extra_values["image"] = ImageAsset(self.image_url)

        if self.manually_approves_followers:
            extra_values["manuallyApprovesFollowers"] = self.manually_approves_followers

        return { **required_document, **extra_values }


src/activity_tools/misc.py
import re

class ImageAsset:

    def __init__(self, url, media_type=None) -> None:
        if not media_type:
            ext = url.split(".")[-1]

            if ext == "jpg" or ext == "jpeg":
                media_type = "image/jpeg"
            elif ext == "png":
                media_type = "image/png"
            else:
                raise Exception(f"Unknown extension {ext}, specify media_type")

        self.media_type = media_type
        self.url = url

    def run(self) -> dict:
        return {
            "type": "Image",
            "mediaType": self.media_type,
            "url": self.url
        }

class Attachment:

    def __init__(self) -> None:
        self.attachments = []

    def add_property_value(self, name, value):
        self.attachments.append({
            "type": "PropertyValue",
            "name": name,
            "value": value
        })

    def run(self) -> dict:
        return self.attachments

class Link:

    def __init__(self, url, text) -> None:
        self.url = url
        self.text = text

    def run(self) -> dict:
        return f'<a href="{self.url}" target="_blank" rel="nofollow noopener noreferrer">{self.text}</a>'


class Tags:

    def __init__(self, domain) -> None:
        self.domain = domain
        self.tags = []

    def add_emoji(self, name, url) -> None:
        self.tags.append({
            "id": f"https://{self.domain}/emoji/{name}",
            "type": "Emoji",
            "name": ":{name}:",
            "icon": ImageAsset(url).run()
        })

    def run(self) -> dict:
        return self.tags

class PublicKey:

    def __init__(self, domain, username, public_key) -> None:
        self.id = f"https://{domain}/users/{username}/key"
        self.owner = f"https://{domain}/users/{username}"
        self.public_key = public_key

    def run(self) -> dict:
        return {
            "id": self.id,
            "owner": self.owner,
            "publicKeyPem": self.public_key,
        }

class WebFinger:

    def __init__(self, domain, resource, actor_url=None) -> None:
        self.domain = domain
        self.resource = resource

        m = re.match(r"^acct:(.+)@(.+)$", resource)
        self.username  = m.group(1)

        if actor_url:
            self.actor_url = actor_url
        else:
            self.actor_url = f"https://{self.domain}/users/{self.username}"

    def run(self) -> dict:
        return {
            "subject": self.resource,
            "links": [
                {
                    "rel": "self",
                    "type": "application/activity+json",
                    "href": self.actor_url
                }
            ]
        }


src/activity_tools/crypto.py
import os

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.backends import default_backend

class RSAKey():
    """
    This class holds private and public RSA keypairs with file
    load and save functionality. This is an utility class that
    you are free to ignore if you prefer to use your own
    cryptographic functions.

    ```python
    Example:
        key = RSAKey("/my/path/key.pem")
        print(key.public_key)
        signed = key.sign(text, padding.PKCS1v15(), hashes.SHA256())
    ```
    """

    def __init__(self, path) -> None:
        """
        This creates/loads/writes a private key to the specified path.
        You can access both the private and public keys from the members
        `.private_key` and `.public_key`.
        """
        self.private_key_path = path

        if os.path.exists(path):
            key = self.load_private_key(path)
        else:
            key = self.generate_private_key()
            self.save_private_key(path)
            print(f"Saved newly generated keys to {path}")

        self.key: rsa.RSAPrivateKey = key
        """ The RSAPrivateKey object from cryptography """

        self.private_key: bytes = self.key.private_bytes(
            serialization.Encoding.PEM,
            serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption())
        """ Contains the private key """

        self.public_key: bytes = self.key.public_key().public_bytes(
            serialization.Encoding.PEM,
            serialization.PublicFormat.SubjectPublicKeyInfo
        )
        """ Contains the public key """

    def generate_private_key(self) -> rsa.RSAPrivateKey:
        """
        Generate a RSA private key and returns it. The constructor calles this
        to generate a private key.
        """
        return rsa.generate_private_key(
            public_exponent=65537, key_size=2048, backend=default_backend()
        )

    def save_private_key(self, path: str) -> None:
        """
        Export the private key to bytes and write them to the specified
        filepath. The constructor uses this to write the keys to storate.
        """
        pem = self.key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption()
        )
        
        with open(path, 'wb') as pem_out:
            pem_out.write(pem)

    def load_private_key(self, path: str) -> rsa.RSAPrivateKey:
        """
        Import and load a private key from a file.
        """
        with open(path, "rb") as key_file:
            key = serialization.load_pem_private_key(
                key_file.read(),
                password=None,
                backend=default_backend()
        )
        return key


src/activity_tools/headers.py
import re
import base64
import hashlib
import json
from typing import Tuple
from datetime import datetime
from urllib.parse import urlparse

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.backends import default_backend

from cryptography.exceptions import InvalidSignature

from .objects import Actor, InboxObject

class ContentTypes:

    @classmethod
    @property
    def activity(cls) -> dict:
        return {
            'Content-Type': "application/activity+json"
        }

    @classmethod
    @property
    def jrd(cls) -> dict:
        return {
            'Content-Type': "application/jrd+json"
        }

class SignatureHeader:

    key_id: str
    algorithm: str
    headers: list
    signature: str

    def __init__(self, header) -> None:
        for sh in header.split(","):
            m = re.match(r'^"?([^"]+)"?="?([^"]+)"?$', sh)
            key = m.group(1)
            value = m.group(2)

            if key.lower() == "keyid":
                self.key_id = value
            elif key.lower() == "algorithm":
                self.algorithm = value
            elif key.lower() == "headers":
                self.headers = value.split(" ")
            elif key.lower() == "signature":
                self.signature = value
            else:
                raise Exception("Unsupported SignatureHeader key")

class Header:
    """
    Representation of a HTTP Header. This class has a bit magic that
    splits and parses headers like the signature header for easy use.
    """

    name: str
    """ The headers name """

    raw_value: str
    """ The unaltered header value """

    _parsed_value: list

    def __init__(self, header) -> None:
        self.name = header[0]
        self.raw_value = header[1]
        self._parsed_value = []

        if self.is_signature():
            self._parsed_value = SignatureHeader(self.raw_value)
        else:
            self._parsed_value = self.raw_value

    @property
    def value(self):
        """ 
        Parsed header value. Most headers are just strings, but
        special headers like the signature header will return a
        SignatureHeader object.
        """
        return self._parsed_value

    def is_signature(self) -> bool:
        """
        Returns true of this is the signature header
        """
        return self.name.lower() == "signature"

class Headers:

    headers: list[Header]

    def __init__(self, headers) -> None:
        self.headers = []

        for header in headers:
            self.headers.append(Header(header))

    def get(self, name):
        for header in self.headers:
            if header.name.lower() == name:
                return header

def verify_signature(
        object: dict,
        headers: list[Tuple[str, str]],
        inbox_path: str
    ) -> bool:

    """
    This function verifies the signature on the specified request. This function
    requires both the incoming object and HTTP headers with the path of the
    receiving inbox. The function will return a boolean value.

    arguments:
    - object     - A dict representing the object we like to verify
    - headers    - A list of tuples of strings representing our HTTP headers
    - inbox_path - The path to our inbox, eg. /users/foo/inbox
    """

    # Analyze headers
    headers_obj = Headers(headers)

    # The extract the value of the signature header
    signature_header: SignatureHeader = headers_obj.get('signature').value

    # Extract the signature, confusually another header with the same name
    # inside the signature header. Decode the signature.
    signature = base64.b64decode(signature_header.signature)

    # Fetch the remote actors, actor
    actor = Actor.fetch(actor_url=object['actor'])

    message = []
    for h in signature_header.headers:
        if h == '(request-target)':
            message.append(f"(request-target): post {inbox_path}")
        else:
            message.append(f"{h}: {headers_obj.get(h).raw_value}")

    message = "\n".join(message).encode("utf-8")

    try:
        actor.get_public_key().verify(
            signature, message, padding.PKCS1v15(), hashes.SHA256()
        )
    except InvalidSignature:
        return False

    return True

def make_signature(remote_inbox: str, message: str, sender_public_key_url: str):
    """
    This function generates a signature for the specified object.
    """

    # The following is to sign the HTTP request as defined in HTTP Signatures.
    private_key_text = open('/tmp/key.pem', 'rb').read() # load from file

    private_key = serialization.load_pem_private_key(
        private_key_text,
        password=None,
        backend=default_backend()
    )

    current_date = datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT')

    recipient_parsed = urlparse(remote_inbox)
    recipient_host = recipient_parsed.netloc
    recipient_path = recipient_parsed.path

    # generating digest
    message_json = json.dumps(message)

    digest = base64.b64encode(
        hashlib.sha256(
            message_json.__str__().encode('utf-8')
        ).digest()
    )

    signature_text = b'(request-target): post %s\ndigest: SHA-256=%s\nhost: %s\ndate: %s' % (
        recipient_path.encode('utf-8'),
        digest,
        recipient_host.encode('utf-8'),
        current_date.encode('utf-8')
    )

    raw_signature = private_key.sign(
        signature_text,
        padding.PKCS1v15(),
        hashes.SHA256()
    )

    signature_header = 'keyId="%s",algorithm="rsa-sha256",headers="(request-target) digest host date",signature="%s"' % (
        sender_public_key_url,
        base64.b64encode(raw_signature).decode('utf-8')
    )

    headers = {
        'Date': current_date,
        'Host': recipient_host,
        'Digest': "SHA-256="+digest.decode('utf-8'),
        'Signature': signature_header
    }

    return headers    


src/activity_tools/objects.py
import requests
import re
import uuid
from urllib.parse import urlparse

from cryptography.hazmat.primitives.serialization import load_pem_public_key

from .misc import ImageAsset, PublicKey, Tags, Attachment

class Actor:
    """
    Generic actor object. Use `create(...)` to create an actor of your own,
    and `fetch(...)` to fetch an external actor.
    """

    username: str
    """ The actors username"""

    def __init__(self) -> None:
        """
        This creates an empty actor object
        """
        self.snake_pattern = re.compile(r'(?<!^)(?=[A-Z])')

    def add_property_value(self, name, value) -> None:
        """
        Add a PropertyValue to the list of attachments. In Mastodon this
        is used for the links in the profile.
        """
        self.attachment.add_property_value(name, value)

    def add_emoji(self, name, url) -> None:
        """
        Add a custom emoji to the tag list. If you for example like to
        map `:foo:` to `/images/foo.png`, this is the function for you!
        """
        self.tag.add_emoji(name, url)

    def create(self, domain: str, username: str, public_key_bytes: bytes) -> None:
        """
        Populate the actor object with data. This is a useful to create the
        actor for a user. All fields can be overriden.
        """
        self.domain = domain
        self.username = username
        self.public_key = None
        self.public_key_pem = public_key_bytes.decode()

        self.id = f"https://{self.domain}/users/{self.username}"
        self.type = "Person"

        self.inbox = f"https://{self.domain}/users/{self.username}/inbox"
        self.outbox = f"https://{self.domain}/users/{self.username}/outbox"

        self.following = f"https://{domain}/users/{username}/following"
        self.followers = f"https://{domain}/users/{username}/followers"

        self.discoverable = False
        self.summary = ""
        self.published = "1523-06-06T10:00:00Z"

        self.name = f"{self.username.capitalize()}"
        self.preferred_username = f"{self.username}"

        self.icon_url = None
        self.image_url = None
        self.manually_approves_followers = None
        self.attachment = Attachment()
        self.tag = Tags(self.domain)

    @classmethod
    def fetch(cls, actor_url):
        actor = Actor()
        actor._fetch(actor_url)
        return actor

    def _fetch(self, actor_url):
        self.actor_raw = {}

        if not actor_url:
            raise Exception("Actor URL is not set")

        headers = {
            'Content-Type': "application/activity+json",
            'Accept': 'application/activity+json'
        }

        actor_resp = requests.get(actor_url, headers=headers)
        if actor_resp.status_code > 299:
            raise Exception(f"Actor {actor_url} responded with a {actor_resp.status_code}")

        self.actor_raw = actor_resp.json()
        urlid = urlparse(self.actor_raw['id'])

        self.domain = urlid.netloc

        keys = [
            "publicKey",
            "id",
            "type",
            "inbox",
            "outbox",
            "following",
            "followers",
            "discoverable",
            "summary",
            "published",
            "name",
            "preferredUsername",
            "icon",
            "image",
            "manuallyApprovesFollowers",
            "attachment",
            "tag"
        ]

        for key in keys:
            key_snake = self.snake_pattern.sub('_', key).lower()
            setattr(self, key_snake, self.actor_raw.get(key))
    
        self.public_key_pem = self.public_key['publicKeyPem']

    def get_public_key(self):
        return load_pem_public_key(self.public_key_pem.encode())

    def run(self) -> dict:
        required_document = {
            "@context": [
                "https://www.w3.org/ns/activitystreams",
                "https://w3id.org/security/v1",
            ],
            "id": self.id,
            "type": self.type,
            "inbox": self.inbox,
            "discoverable": self.discoverable,
            "summary": self.summary,
            "published": self.published,
            "name": self.name,
            "preferredUsername": self.preferred_username,
            "attachment": self.attachment.run(),
            "tag": self.tag.run(),
            "publicKey": PublicKey(
                self.domain,
                self.username,
                self.public_key_pem
            ).run(),
        }

        extra_values = {}

        if self.icon_url:
            extra_values["icon"] = ImageAsset(self.icon_url)

        if self.icon_url:
            extra_values["image"] = ImageAsset(self.image_url)

        if self.manually_approves_followers:
            extra_values["manuallyApprovesFollowers"] = self.manually_approves_followers

        if self.followers:
            extra_values["followers"] = self.followers

        if self.following:
            extra_values["following"] = self.following

        if self.outbox:
            extra_values["outbox"] = self.outbox

        return { **required_document, **extra_values }


class WrapActivityStreamsObject:

    def __init__(self, object) -> None:
        self.object = object

    def run(self) -> dict:
        context = {
            "@context": "https://www.w3.org/ns/activitystreams"
        }

        return { **context, **self.object.run() }

class InboxObject:

    raw: dict
    """
    The raw data representing this object
    """

    def __init__(self, data) -> None:
        self.raw = data
        self.id = data['id']
        self.type = data['type'].lower()
        self._actor = data['actor']
        self._object = data['object']

    @property
    def actor(self) -> Actor:
        return Actor.fetch(actor_url=self._actor)

    @property
    def object(self) -> Actor:
        return Actor.fetch(actor_url=self._object)

    def run(self) -> dict:
        """
        Return the raw JSON data
        """
        return self.raw

class Follow(InboxObject):

    def __init__(self, data) -> None:
        super().__init__(data)

class Undo(InboxObject):

    def __init__(self, data) -> None:
        super().__init__(data)

class ActivityPubObject:
    
    id: str
    """
    The objects ID, this should be a unique identifier
    """

    actor: Actor
    """
    The actor that created the message (the sender)
    """

    actor_url: str
    """
    A url to the actor that created the message (the sender)
    """

    object: object
    """
    A generic object representing the message/data. This may also
    point to a target actor.
    """

    raw: dict
    """
    Object in raw unaltered form represented as a dict. This may be the
    unaltered JSON document, but not always.
    """

class Accept(ActivityPubObject):

    object: Follow
    """ Object representing the Follow request """

    def __init__(self, object: InboxObject) -> None:
        self.id = f"{object.object.id}/accept/{str(uuid.uuid4())}"
        self.object = object
        self.raw = object.run()
        self.actor = object.object

    def run(self):
        return {
            "@context": "https://www.w3.org/ns/activitystreams",
            "id": self.id,
            "type": "Accept",
            "actor": self.object.object.id,
            "object": self.object.run()
        }
